File: packages/neovim-ide-companion/lua/gemini-cli/discovery.lua
-- Copyright 2026 Google LLC
-- SPDX-License-Identifier: Apache-2.0

local M = {}

--- @type string|nil Path to the discovery file
local discovery_file = nil

--- Get the temp directory.
--- @return string
local function get_tmpdir()
  return os.getenv("TMPDIR") or os.getenv("TEMP") or os.getenv("TMP") or "/tmp"
end

--- Get the discovery directory path.
--- @return string
local function get_discovery_dir()
  local tmpdir = get_tmpdir()
  return tmpdir .. "/gemini/ide"
end

--- Write the discovery file.
--- @param port number
--- @param auth_token string
--- @param workspace_path string|nil
--- @return string|nil file_path, string|nil error
function M.write(port, auth_token, workspace_path)
  local dir = get_discovery_dir()
  local pid = vim.fn.getpid()
  local filename = string.format("gemini-ide-server-%d-%d.json", pid, port)

  -- Create directory recursively
  vim.fn.mkdir(dir, "p")

  local filepath = dir .. "/" .. filename

  local content = vim.json.encode({
    port = port,
    workspacePath = workspace_path or vim.fn.getcwd(),
    authToken = auth_token,
    ideInfo = {
      name = "neovim",
      displayName = "Neovim",
    },
  })

  local f, err = io.open(filepath, "w")
  if not f then
    return nil, "Failed to write discovery file: " .. (err or "unknown error")
  end
  f:write(content)
  f:close()

  -- Set file permissions to 0600
  local uv = vim.loop
  uv.fs_chmod(filepath, tonumber("600", 8))

  discovery_file = filepath
  return filepath, nil
end

--- Remove the discovery file.
function M.cleanup()
  if discovery_file then
    pcall(os.remove, discovery_file)
    discovery_file = nil
  end
end

--- Get the current discovery file path.
--- @return string|nil
function M.get_path()
  return discovery_file
end

return M

File: packages/zed-ide-companion/src/discovery.rs
// Copyright 2026 Google LLC
// SPDX-License-Identifier: Apache-2.0

use serde_json::json;
use std::fs;
use std::os::unix::fs::PermissionsExt;
use std::path::PathBuf;

/// Write the discovery file and return its path.
pub fn write(port: u16, auth_token: &str, workspace_path: &str) -> Result<PathBuf, String> {
    let tmpdir = std::env::var("TMPDIR")
        .or_else(|_| std::env::var("TEMP"))
        .or_else(|_| std::env::var("TMP"))
        .unwrap_or_else(|_| "/tmp".to_string());

    let dir = PathBuf::from(&tmpdir).join("gemini").join("ide");
    fs::create_dir_all(&dir).map_err(|e| format!("Failed to create discovery dir: {}", e))?;

    let pid = std::process::id();
    let filename = format!("gemini-ide-server-{}-{}.json", pid, port);
    let filepath = dir.join(&filename);

    let content = json!({
        "port": port,
        "workspacePath": workspace_path,
        "authToken": auth_token,
        "ideInfo": {
            "name": "zed",
            "displayName": "Zed"
        }
    });

    fs::write(&filepath, content.to_string())
        .map_err(|e| format!("Failed to write discovery file: {}", e))?;

    // Set permissions to 0600
    fs::set_permissions(&filepath, fs::Permissions::from_mode(0o600))
        .map_err(|e| format!("Failed to set permissions: {}", e))?;

    Ok(filepath)
}

/// Remove the discovery file.
pub fn cleanup(path: &PathBuf) {
    let _ = fs::remove_file(path);
}
